docs: security hardening info for actions untrusted content (#38048)
github/docs: content/actions/reference/security/secure-use.md

To help mitigate the risk of an exposed token, consider restricting the assigned permissions. For more information, see [AUTOTITLE](/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token).

{% ifversion custom-org-roles %}

> ## Mitigating the risks of untrusted code checkout

Reviewer: This new H2 and the H3 below it look good!

> ## Understanding the risks of untrusted code checkout

Reviewer 2: ```suggestion
## Mitigating the risks of untrusted code checkout
```

> Similar to script injection attacks, untrusted pull request content that automatically triggers actions processing can also pose a security risk. The `pull_request_target` and `workflow_run` workflow triggers, when used with the checkout of an untrusted pull request, expose the repository to security compromises. These workflows are privileged (i.e. they share the same cache of the main branch with other privileged workflow triggers, may have repository write access and access to referenced secrets), which can be exploited to take over a repository.

Reviewer 2: ```suggestion
Similar to script injection attacks, untrusted pull request content that automatically triggers actions processing can also pose a security risk. The `pull_request_target` and `workflow_run` workflow triggers, when used with the checkout of an untrusted pull request, expose the repository to security compromises. These workflows are privileged, which means they share the same cache of the main branch with other privileged workflow triggers, and may have repository write access and access to referenced secrets. These vulnerabilities can be exploited to take over a repository.
```

Author: Yeah, this is nice and clear to me.

> For more information on these triggers, how to use them, and the associated risks, see [AUTOTITLE](/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request_target) and the [AUTOTITLE](/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#workflow-run).

Reviewer: ```suggestion
For more information on these triggers, how to use them, and the associated risks, see [AUTOTITLE](/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request_target) and [AUTOTITLE](/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#workflow_run).
```

V minor (unnecessary word and broken link).

> Review the [`pull_request_target` trigger documentation](/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request_target) and the [`workflow_run` trigger documentation](/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#workflow-run) for more information on these triggers, how to use them, and the risks associated with them.

Reviewer 2: ```suggestion
For more information on these triggers, how to use them, and the associated risks, see [AUTOTITLE](/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request_target) and the [AUTOTITLE](/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#workflow-run).
```

For additional explanation, examples, and guidance on the risks of untrusted code checkout, see [Keeping your {% data variables.product.prodname_actions %} and workflows secure: Preventing pwn requests](https://securitylab.github.com/research/github-actions-preventing-pwn-requests/) from {% data variables.product.prodname_security %} and the [Dangerous Workflow check documentation](https://github.com/ossf/scorecard/blob/main/docs/checks.md#dangerous-workflow) from OpenSSF Scorecard.

### Good practices

There are a number of different approaches available to help you mitigate the risk of untrusted code checkout in an action workflow:

* Avoid using the `pull_request_target` workflow trigger if not necessary. Prefer using `workflow_run` for privilege separation between workflows as described in [Preventing pwn requests](https://securitylab.github.com/research/github-actions-preventing-pwn-requests). Only use these workflow triggers when the workflow actually needs the privileged context.

> * Avoid using the `pull_request_target` and `workflow_run` workflow triggers with untrusted pull requests or code content. Workflows that use these triggers must not explicitly checkout untrusted code, including from pull request forks or from repositories that are not under your control. Workflows triggered on `workflow_run` should treat artifacts uploaded from other workflows with caution.

Reviewer: Good call—in this case it's verbal, so this suggestion is correct. We don't always get this right across our docs.

> * Avoid using the `pull_request_target` and `workflow_run` workflow triggers with untrusted pull requests or code content. Workflows that use these triggers must not explicitly checkout untrusted code, including from pull request forks or from repositories that are not under your control. Workflows triggered on `workflow_run` should treat artifacts uploaded from other workflows with caution.

Author: In this context I expected "checkout" to be a technical term (`git checkout`). Maybe a clearer phrase would be "use or execute" to cover any user action that causes untrusted code to be used.

* [OpenSSF Scorecards](#use-openssf-scorecards-to-detect-potentially-vulnerable-workflows) can help you identify potentially vulnerable workflows, along with other security risks when using {% data variables.product.prodname_actions %}.

## Managing permissions for {% data variables.product.prodname_actions %} settings in your organization

You can grant an individual or team in your organization the ability to control certain subsets of settings without granting full administrative control of the organization and its repositories.

{% data reusables.actions.org-roles-for-gh-actions %}

For more information, see [AUTOTITLE](/organizations/managing-peoples-access-to-your-organization-with-roles/about-custom-organization-roles).

{% endif %}

## Using OpenID Connect to access cloud resources

{% data reusables.actions.about-oidc-short-overview %}

> {% data reusables.actions.oidc-custom-claims-aws-restriction %}

Reviewer: @wrslatz I realise we've spent some time refining the suggested wording here, so first of all apologies for the confusion.

I'm just unsure why we'd add these two sections, since we already have them on the page as H2s, so my understanding is that they'd apply in a general sense to best practices for using GitHub Actions and its features? That is, not specifically for "Mitigating the risks of untrusted code checkout" but still relevant to that.

For clarity: I'm referring to lines 130-144 at time of writing.

Author: I think these might be unintentional additions or changes caused by my changes going in near them. Looking back at old commits, I think these were already there and moved around and that's why they're showing? If we confirm that's the case and the content has since moved elsewhere, we can remove them from here.

Reviewer 2: Yeah, lines 129-144 all got duplicated, so I'm going to get rid of those.

There are a number of different approaches available to help you mitigate the risk of untrusted code checkout in an action workflow:

### Avoid potentially dangerous workflow triggers

Avoid using the `pull_request_target` workflow trigger if not necessary. Prefer using `workflow_run` for privilege separation between workflows as described in [Keeping your {% data variables.product.prodname_actions %} and workflows secure: Preventing pwn requests](https://securitylab.github.com/research/github-actions-preventing-pwn-requests). Only use these workflow triggers when the workflow actually needs the privileged context.

### Do not use the `pull_request_target` and `workflow_run` workflow triggers with untrusted content

Avoid using the `pull_request_target` and `workflow_run` workflow triggers with untrusted pull requests or code content. Workflows that use these triggers must not explicitly checkout untrusted code, including from pull request forks or from repositories that are not under your control. Workflows triggered on `workflow_run` should treat artifacts uploaded from other workflows with caution (i.e. as untrusted).

> ### Use CodeQL to detect potentially vulnerable workflows

Reviewer 2: ```suggestion
### Use {% data variables.product.prodname_codeql %} to detect potentially vulnerable workflows
```

> [CodeQL](/code-security/code-scanning/introduction-to-code-scanning/about-code-scanning-with-codeql) can scan and detect potentially vulnerable {% data variables.product.prodname_actions %} workflows. [Configure the default setup for CodeQL](/code-security/code-scanning/enabling-code-scanning/configuring-default-setup-for-code-scanning) for the repository and ensure that {% data variables.product.prodname_actions %} scanning is enabled.

Reviewer 2: ```suggestion
[{% data variables.product.prodname_codeql %}](/code-security/code-scanning/introduction-to-code-scanning/about-code-scanning-with-codeql) can scan and detect potentially vulnerable {% data variables.product.prodname_actions %} workflows. [Configure the default setup for {% data variables.product.prodname_codeql %}](/code-security/code-scanning/enabling-code-scanning/configuring-default-setup-for-code-scanning) for the repository and ensure that {% data variables.product.prodname_actions %} scanning is enabled.
```

### Use OpenSSF Scorecards to detect potentially vulnerable workflows

[OpenSSF Scorecards](#use-openssf-scorecards-to-detect-potentially-vulnerable-workflows) can help you identify potentially vulnerable workflows, along with other security risks when using {% data variables.product.prodname_actions %}.

## Managing permissions for {% data variables.product.prodname_actions %} settings in your organization

> You can practice the principle of least privilege for your organization's CI/CD pipeline with {% data variables.product.prodname_actions %} by administering custom organization roles. A custom organization role is a way to grant an individual or team in your organization the ability to control certain subsets of settings without granting full administrative control of the organization and its repositories.

Reviewer 2: ```suggestion
You can grant an individual or team in your organization the ability to control certain subsets of settings without granting full administrative control of the organization and its repositories.
```
